bundles/org.openhab.binding.ecotouch/src/main/java/org/openhab/binding/ecotouch/internal/EcoTouchConnector.java

```diff
@@ -0,0 +1,330 @@
+                logouturl.openConnection();
+            } catch (Exception e) {
+            }
+            cookies = null;
+        }
+    }
+
+    /**
+     * Request a value from the heat pump
+     * 
+     * @param tag
+     *            The register to query (e.g. "A1")
+     * @return value This value is a 16-bit integer.
+     */
+    public String getValue(String tag) throws Exception {
+        Map<String, String> result = getValues(Set.of(tag));
+        String value = result.get(tag);
+        if (value == null) {
+            // failed
+            logger.debug("Cannot get value for tag '{}' from Waterkotte EcoTouch.", tag);
+            throw new EcoTouchException("invalid response from EcoTouch");
+        }
+
+        return value;
+    }
+
+    /**
+     * Request multiple values from the heat pump
+     * 
+     * @param tags
+     *            The registers to query (e.g. "A1")
+     * @return values A map of tags and their respective string values
+     */
+    public Map<String, String> getValues(Set<String> tags) throws Exception {
+        final Integer maxNum = 100;
+
+        Map<String, String> result = new HashMap<String, String>();
+        Integer counter = 1;
+        StringBuilder query = new StringBuilder();
+        Iterator<String> iter = tags.iterator();
+        while (iter.hasNext()) {
+            query.append(String.format("t%d=%s&", counter, iter.next()));
+            counter++;
+            if (counter > maxNum) {
+                query.deleteCharAt(query.length() - 1); // remove last '&'
+                String queryStr = String.format("http://%s/cgi/readTags?n=%d&", ip, maxNum) + query;
+                result.putAll(getValues(queryStr));
+                counter = 1;
+                query = new StringBuilder();
+            }
+        }
+
+        if (query.length() > 0) {
+            query.deleteCharAt(query.length() - 1); // remove last '&'
+            String queryStr = String.format("http://%s/cgi/readTags?n=%d&", ip, counter - 1) + query;
+            result.putAll(getValues(queryStr));
+        }
+
+        return result;
+    }
+
+    /**
+     * Send a request to the heat pump and evaluate the result
+     * 
+     * @param url
+     *            The URL to connect to
+     * @return values A map of tags and their respective string values
+     */
+    private Map<String, String> getValues(String url) throws Exception {
+        trylogin(false);
+        Map<String, String> result = new HashMap<String, String>();
+        int loginAttempt = 0;
+        while (loginAttempt < 2) {
+            BufferedReader reader = null;
+            try {
+                URLConnection connection = new URL(url).openConnection();
+                if (cookies != null) {
+                    for (String cookie : cookies) {
+                        connection.addRequestProperty("Cookie", cookie.split(";", 2)[0]);
+                    }
+                }
```

Comment:
Here it get the following warning:
`[WARNING] org.openhab.binding.ecotouch/src/main/java/org/openhab/binding/ecotouch/internal/EcoTouchConnector.java:[201,42] Potential null pointer access: this expression has a '@Nullable' type`
cookies is Nullable, but I check for null before use. I'm lost here.

Reply:
```suggestion
List<String> localCookies=cookies;
                if (localCookies != null) {
                    for (String cookie : localCookies) {
                        connection.addRequestProperty("Cookie", cookie.split(";", 2)[0]);
                    }
                }
```

Reply:
does that mean, that the same class instance is executed by two different tasks at the same time? It is clear, that I can have more than one instance, if I have more than one heat pump. I thought serialization is performed by the framework.
If that is not true, I'll have to check all places for data races and have mutexes all around... That seems unreasonable and unlikely to me...
I would guess that the static checker is just not aware of the framework's serialization. Isn't there a magic attribute to provide that info?

Reply:
I am not the best person to explain things to you as I am new to Java. Does the code suggested solve the compiler's warning for null? Your concerns about data races and serialization are not related to this, it is about potentially another thread making the handle to the object null if my understanding is correct.

Reply:
The root of the problem is: the compiler thinks cookies can turn into null between line 200 and 201. That's only possible if multithreading is allowed AND the very same instance of the object is executed in two different threads. Your workaround would prevent the compiler from flagging the line 201 as a problem, because you created a local variable on the stack, which is bound to the thread. This will "fix" this particular problem (with an ugly workaround) but if the compiler is right in flagging it, then I have to check the complete code for data races.
Is there any documentation, how openhab implements multithreading? Again, I think the compiler warning is a false positive, because otherwise it would make binding implementation unnecessary complicated.

Reply:
Yes it is a trade off, so where it is possible variables should be nonNullable and initialized at creation. Whilst it may be that you know your code 100% inside out, if someone else comes along to maintain it they may not know it as well. My knowledge is not very good, if you want to discuss how openHAB implements multithreading you should ask on the forum in the dev section. The suggestion I made to you, is what others have given me when I needed help.